src/MassTransit/Transports/ConsumerAgent.cs

```diff
@@ -0,0 +1,176 @@
+
+                _deliveryComplete.TrySetResult(true);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected void TrySetConsumeTask()
+        {
+            if (_consumeTask != null || _consumeTaskSource.IsValueCreated)
+                return;
+
+            lock (_lock)
+            {
+                if (_consumeTask != null || _consumeTaskSource.IsValueCreated)
+                    return;
+
+                SetConsumeTask(_consumeTaskSource.Value.Task);
+            }
+        }
+
+        protected void TrySetConsumeTask(Task consumeTask)
+        {
+            if (_consumeTask != null)
+                return;
+
+            lock (_lock)
+            {
+                if (_consumeTask != null)
+                    return;
+
+                _consumeTask = consumeTask;
+                SetConsumeTask(_consumeTask);
+            }
+        }
+
+        void SetConsumeTask(Task consumeTask)
+        {
+            async Task TryStop()
+            {
+                if (IsStopping)
+                    return;
+
+                IsGracefulShutdown = false;
+
+                try
+                {
+                    LogContext.SetCurrentIfNull(_context.LogContext);
+
+                    await this.Stop("Consume Loop Exited").ConfigureAwait(false);
+                }
+                catch (Exception exception)
+                {
+                    LogContext.Warning?.Log(exception, "Stop Faulted");
+                }
+            }
+
+            if (consumeTask.IsCompleted)
+                Task.Run(() => TryStop());
+            else
+                consumeTask.ContinueWith(_ => TryStop(), TaskContinuationOptions.RunContinuationsAsynchronously);
+        }
+
+        protected override Task StopAgent(StopContext context)
+        {
+            LogContext.Debug?.Log("Stopping consumer: {InputAddress} with a reason: {Reason}", _context.InputAddress, context.Reason);
+
+            if (_consumeTaskSource.IsValueCreated)
+                _consumeTaskSource.Value.TrySetResult(true);
+
+            SetCompleted(ActiveAndActualAgentsCompleted(context));
+
+            return Completed;
+        }
+
+        protected void TrySetConsumeCanceled(CancellationToken cancellationToken = default)
+        {
+            if (!_consumeTaskSource.IsValueCreated)
+                return;
+
+            if (ActiveDispatchCount == 0)
+                _deliveryComplete.TrySetResult(false);
+
+            _consumeTaskSource.Value.TrySetCanceled(cancellationToken);
+        }
+
+        protected void TrySetConsumeException(Exception exception)
+        {
+            if (!_consumeTaskSource.IsValueCreated)
+                return;
+
+            _deliveryComplete.TrySetResult(false);
+
+            _consumeTaskSource.Value.TrySetException(exception);
+        }
+
+        protected virtual async Task ActiveAndActualAgentsCompleted(StopContext context)
```

Comment:
Seems like this should be internalized, and should somehow call another method for the subclass to close/etc. Like ActiveMQ overrides, but doesn't call base, so this logic is never executed.